### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by Melodicolt**
- feat: exclude models in testing and add logging for coverage to phpun… …
- fix: add reason to rejected conversion
- test: improve tests for bundle, full coverage

### CODE REVIEW

File: tests/Client/PartnerizeClientTest.php [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+            ->method('deserialize')
+            ->willReturn($partnerizeResponse);
+
+        $client = new PartnerizeClient($this->trackingClient, $this->apiClient, $this->campaignId, $this->serializer);
+        $job = $client->approveConversion('testId');
+
+        $this->assertEquals($expectedJob, $job);
+    }
+
+    public function testJobReturnedOnReject(): void
```
Comment (kiettran) [inline tests/Client/PartnerizeClientTest.php]: Can you explain what you are testing by filling up the docblock?

File: tests/Client/PartnerizeClientTest.php [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+            ->willReturn($response);
+
+        $conversionId = $this->partnerizeClient->createConversion($sale);
+        $this->assertEquals('AShinyNewConversionId', $conversionId);
+    }
+
+    /**
+     * Test that the client throws client exceptions and doesnt "leak" guzzle exceptions.
+     */
+    public function testCreateConversionError(): void
```
Comment (kiettran) [inline tests/Client/PartnerizeClientTest.php]: Perhaps is `testCreateConversionWillThrowAClientException` a better name for this test.

**Review by kiettran — COMMENTED** [review]
If I am correct, you have three different functions within this class.
- createConversion
- approveConversion
- rejectConversion

One function does a get request and the other two functions uses the private function setConversionStatus where the post request happens and may throw a `ClientException`.

What if you split them up in the following tests?
- testCreateConversion
- testCreateConversionReturnsEmptyResponse
- testCreateConversionWillThrowClientException
- testApproveConversion
- testApproveConversionWillThrowClientException
- testApproveConversionReceivesBadStatusCodeWillThrowClientException
- testRejectConversion
- testRejectConversionWillThrowClientException
- testRejectConversionReceivesBadStatusCodeWillThrowClientException

Then you'll have a better view of what you're testing in your class.

File: tests/Client/PartnerizeClientTest.php (hunk 2 above)
Comment (Melodicolt) [reply tests/Client/PartnerizeClientTest.php]: Restructured the tests.

File: tests/Client/PartnerizeClientTest.php (hunk 1 above)
Comment (Melodicolt) [reply tests/Client/PartnerizeClientTest.php]: Restructured the tests.

File: tests/Encoder/PartnerizeS2SEncoderTest.php [hunk 3]
```diff
@@ -12,15 +12,87 @@
...
+    public function testSupportsEncodingReturnsBoolean($encoding, $expected): void
+    {
+        $result = $this->encoder->supportsEncoding($encoding);
+        $this->assertEquals($expected, $result);
+    }
+
+    /**
+     * Test that the encoder throws a RuntimeException on bad input data
+     */
+    public function testRuntimeExceptionIfDataNotArrayOnEncoding(): void
```
Comment (kiettran) [inline tests/Encoder/PartnerizeS2SEncoderTest.php]: What if you rename this test function to `testEncodeWillThrowRuntimeExceptionWhenDataIsNotAnArray`? Then you'll read that you're actually testing the Encode function but will throw an exception when the given data is not an array.

File: tests/Encoder/PartnerizeS2SEncoderTest.php [hunk 4]
```diff
@@ -12,15 +12,87 @@
...
+        $this->expectException(\RuntimeException::class);
+        $this->expectExceptionMessage('Expected first parameter to be an array, but got string.');
+
+        $this->encoder->encode('test', 'json');
+    }
+
+    /**
+     * Test that the encoder throws a RuntimeException on unnormalized data
+     */
+    public function testRuntimeExceptionIfValueNotScalarOnEncoding(): void
```
Comment (kiettran) [inline tests/Encoder/PartnerizeS2SEncoderTest.php]: Same here, what if you rename this test function to `testEncodeWillThrowRuntimeExceptionWhenDataValueIsNotScalar`? Then you'll read that you're actually testing the Encode function but will throw an exception when the given data value is not scalar.

**Review by kiettran — APPROVED** [review]
Some minor changes about naming the test functions, but overall good job!

File: tests/Encoder/PartnerizeS2SEncoderTest.php (hunk 3 above)
Comment (Melodicolt) [reply tests/Encoder/PartnerizeS2SEncoderTest.php]: Refactored tests according to above feedback.

File: tests/Encoder/PartnerizeS2SEncoderTest.php (hunk 4 above)
Comment (Melodicolt) [reply tests/Encoder/PartnerizeS2SEncoderTest.php]: Refactored tests according to above feedback.

**Comment by Melodicolt** [comment]
Just a reminder that this still needs to be approved, if everything is A okay!

**Review by stixx — APPROVED** [review]

**Review by joostvanveelen — APPROVED** [review]

**Pull request merged**